File: WebClient SDK/src/default/ui/krProcess/krTilesUIExtension.ts
import { runInAction } from 'mobx';
import { KrTileInflater } from '../../workflow/krProcess/krTileInflater';
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { ITilePanel, Tile, TileGroups, ITile } from 'tessa/ui/tiles';
import { cardInSpecialMode, tryGetFromInfo } from 'tessa/ui';
import { KrTileInfo } from 'tessa/workflow/krProcess';
import { IStorage } from 'tessa/platform/storage';

export class KrTilesUIExtension extends CardUIExtension {

  public contextInitialized(context: ICardUIExtensionContext) {
    const card = context.card;
    let tilesPanel: ITilePanel;
    if (!card
      || cardInSpecialMode(context.model)
      || !context.uiContext.tiles
      || !(tilesPanel = context.uiContext.tiles.leftPanel)
    ) {
      return;
    }

    const tilesStorage = tryGetFromInfo<IStorage | null>(card.info, 'LocalTilesInfoMark', null);
    if (!tilesStorage) {
      return;
    }

    const localTiles: KrTileInfo[] = [];
    if (Array.isArray(tilesStorage)) {
      for (let tile of tilesStorage) {
        localTiles.push(new KrTileInfo(tile));
      }
    } else {
      const names = Object.getOwnPropertyNames(tilesStorage);
      for (let name of names) {
        localTiles.push(new KrTileInfo(tilesStorage[name]));
      }
    }

    // т.к. в этот момент tilesPanel.tiles уже observable, то нужно завернуть в runInAction
    runInAction(() => {
      KrTilesUIExtension.removeTilesWithTileInfo(tilesPanel.tiles);

      const tiles = KrTileInflater.instance.inflate(tilesPanel.contextSource, localTiles, TileGroups.Workflow);

      const actionGroupingTiles = tiles
        .filter(x => {
          const info = tryGetFromInfo<KrTileInfo | null>(x.sharedInfo, 'TileInfo', null);
          return info && info.actionGrouping
        })
        .map(x => x.clone());

      let groupTile = tilesPanel.tryGetTile('ActionsGrouping');
      if (groupTile) {
        KrTilesUIExtension.removeTilesWithTileInfo(groupTile.tiles);
      }

      if (actionGroupingTiles.length > 0) {
        if (!groupTile) {
          groupTile = new Tile({
            name: 'ActionsGrouping',
            caption: '$UI_Tiles_ActionsGrouping',
            icon: 'ta icon-thin-258',
            contextSource: tilesPanel.contextSource,
            group: TileGroups.Cards,
            order: 10
          });
          tilesPanel.tiles.push(groupTile);
        }

        groupTile.info['MinActionsGroupingCount'] = 1;
        groupTile.tiles.push(...actionGroupingTiles);
      }

      // TODO: может быть сюда надо пихать тайлы с фильтром !actionGrouping
      tilesPanel.tiles.push(...tiles);
    });
  }

  private static removeTilesWithTileInfo(tiles: ITile[]) {
    const removeIndices: number[] = [];
    tiles.forEach((x, i) => {
      if (tryGetFromInfo(x.sharedInfo, 'TileInfo', null)) {
        removeIndices.push(i);
      }
    });
    removeIndices.reverse();

    for (let i of removeIndices) {
      tiles.splice(i, 1);
    }
  }

}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/KrTilesUIExtension.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Input;
using Tessa.Cards;
using Tessa.Extensions.Default.Client.Workflow.KrProcess;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform;
using Tessa.Platform.Collections;
using Tessa.Platform.Settings;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;

namespace Tessa.Extensions.Default.Client.UI.KrProcess
{
    public sealed class KrTilesUIExtension :
        CardUIExtension
    {
        #region Fields

        private readonly IKrTileInflater tileInflater;

        private const string tileCommandKey = CardHelper.SystemKeyPrefix + "TileCommand";
        private const string gestureKey = CardHelper.SystemKeyPrefix + "Gesture";

        #endregion

        #region Constructors

        public KrTilesUIExtension(
            IKrTileInflater tileInflater) => this.tileInflater = tileInflater;

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override async Task ContextInitialized(
            ICardUIExtensionContext context)
        {
            var card = context.Card;
            ITilePanel tilesPanel;
            if (card is null
                || context.Model.Flags.InSpecialMode()
                || (tilesPanel = context.UIContext.Tiles?.LeftPanel) is null)
            {
                return;
            }

            tilesPanel.Tiles.RemoveAll(static p =>
            {
                if (p.SharedInfo.ContainsKey(KrConstants.Ui.TileInfo))
                {
                    RemoveHotkey(p);
                    return true;
                }
                return false;
            });

            if (!card.TryGetLocalTiles(out var tileInfos))
            {
                return;
            }

            var tiles = this.tileInflater.Inflate(tilesPanel, tileInfos);

            var actionGroupingTiles = tiles
                .Where(static p =>
                    p.SharedInfo.TryGet<KrTileInfo>(KrConstants.Ui.TileInfo)?.ActionGrouping == true)
                .Select(static p => p.Clone())
                .ToList();

            if (actionGroupingTiles.Count > 0)
            {
                var groupTile = tilesPanel.Tiles.TryGet(TileNames.ActionsGrouping);
                if (groupTile is null)
                {
                    groupTile = new Tile(
                        TileNames.ActionsGrouping,
                        "$UI_Tiles_ActionsGrouping",
                        context.Icons.Get("Thin258"),
                        tilesPanel,
                        DelegateCommand.Empty,
                        TileGroups.Cards,
                        order: 10);
                    tilesPanel.Tiles.Add(groupTile);
                }
                groupTile.Tiles.RemoveAll(static p =>
                    p.SharedInfo.ContainsKey(KrConstants.Ui.TileInfo));

                groupTile.Info[nameof(ISettings.MinActionsGroupingCount)] = Int32Boxes.One;
                groupTile.Tiles.AddRange(actionGroupingTiles);
            }

            tilesPanel.Tiles.AddRange(tiles);

            SetHotkeys(tiles, context);
        }

        #endregion

        #region Private Methods

        private static void SetHotkeys(
            ICollection<ITile> tiles,
            ICardUIExtensionContext context)
        {
            var keyGestureConverter = (TypeConverter) new KeyGestureConverter();
            foreach (var tile in tiles)
            {
                if (tile.Tiles is not null
                    && tile.Tiles.Count > 0)
                {
                    SetHotkeys(tile.Tiles, context);
                }

                var tileInfo = tile.SharedInfo.TryGet<KrTileInfo>(KrConstants.Ui.TileInfo);
                if (tileInfo is not null
                    && !string.IsNullOrWhiteSpace(tileInfo.ButtonHotkey))
                {
                    try
                    {
                        var gesture = (KeyGesture) keyGestureConverter.ConvertFromString(tileInfo.ButtonHotkey);
                        var tileCommand = new DelegateCommand(async p =>
                            await tile.ExecuteCommandWithCheckAsync(TileCommandEventType.KeyPressed));

                        tile.Context.InputBindings.AddCombined(
                            tileCommand,
                            gesture);

                        tile.Info[tileCommandKey] = tileCommand;
                        tile.Info[gestureKey] = gesture;
                    }
                    catch (ArgumentException e)
                    {
                        ValidationSequence
                            .Begin(context.ValidationResult)
                            .SetObjectName(nameof(KrTilesUIExtension))
                            .WarningException(
                                e,
                                $"Invalid hotkey {tileInfo.ButtonHotkey} in KrSecondaryProcess {tileInfo.Name} (ID = {tileInfo.ID}).")
                            .End();
                    }
                }
            }
        }

        private static void RemoveHotkey(ITile tile)
        {
            if (tile.Info.TryGetValue(tileCommandKey, out var tileCommandObj)
                && tile.Info.TryGetValue(gestureKey, out var gestureObj))
            {
                tile.Context.InputBindings.RemoveCombined(
                    (ICommand) tileCommandObj,
                    (InputGesture) gestureObj,
                    static (c1, c2) => c1 == c2);
            }

            if (tile.Tiles is not null
                && tile.Tiles.Count > 0)
            {
                foreach (var innerTile in tile.Tiles)
                {
                    RemoveHotkey(innerTile);
                }
            }
        }

        #endregion

    }
}
